solver/src/solver/matcha_solver.rs

```diff
@@ -0,0 +1,312 @@
+//! Module containing implementation of the Matcha solver.
+//!
+//! This simple solver will simply use the Matcha API to get a quote for a
+//! single GPv2 order and produce a settlement directly against Matcha.
+
+pub mod api;
+
+use super::solver_utils::{AllowanceFetching, Slippage};
+use crate::solver::matcha_solver::api::MatchaApi;
+use anyhow::{ensure, Result};
+use contracts::{GPv2Settlement, ERC20};
+use ethcontract::{dyns::DynWeb3, Bytes, U256};
+use maplit::hashmap;
+
+use super::single_order_solver::SingleOrderSolving;
+
+use self::api::{DefaultMatchaApi, SwapQuery, SwapResponse};
+use crate::{
+    encoding::EncodedInteraction,
+    interactions::Erc20ApproveInteraction,
+    liquidity::LimitOrder,
+    settlement::{Interaction, Settlement},
+};
+use model::order::OrderKind;
+use std::fmt::{self, Display, Formatter};
+
+/// Constant maximum slippage of 5 BPS (0.05%) to use for on-chain liquidity.
+/// This is halve the 1inch slippage
+pub const STANDARD_MATCHA_SLIPPAGE_BPS: u16 = 5;
+
+/// A GPv2 solver that matches GP orders to direct Matcha swaps.
+pub struct MatchaSolver<F> {
+    settlement_contract: GPv2Settlement,
+    client: Box<dyn MatchaApi + Send + Sync>,
+    allowance_fetcher: F,
+}
+
+/// Chain ID for Mainnet.
+const MAINNET_CHAIN_ID: u64 = 1;
+
+impl MatchaSolver<GPv2Settlement> {
+    pub fn new(settlement_contract: GPv2Settlement, chain_id: u64) -> Result<Self> {
+        ensure!(
+            chain_id == MAINNET_CHAIN_ID,
+            "Matcha solver only supported on Mainnet",
+        );
+
+        let allowance_fetcher = settlement_contract.clone();
+        Ok(Self {
+            settlement_contract,
+            allowance_fetcher,
+            client: Box::new(DefaultMatchaApi::default()),
+        })
+    }
+}
+impl<F> MatchaSolver<F> {
+    fn web3(&self) -> DynWeb3 {
+        self.settlement_contract.raw_instance().web3()
+    }
+}
+
+impl<F> std::fmt::Debug for MatchaSolver<F> {
```

Comment:
I see that we have implemented display and debug both the same for this solver. Makes me wonder if we could just get away with one. Is there anything about using display for debug if display is defined. Like a fallback for the better one in the hierarchy.